README.MD
# NeuroElf (MATLAB)

Please note: if you clone/download from GitHub or install from a "flat" folder ZIP file, you must manually add the NeuroElf main folder to the MATLAB path (without sub-folders), and subsequently call the `neuroelf_setup` function:

```
>> neuroelf_setup;
```

and then follow the prompts on screen, to select a suitable temporary folder and decide whether or not you wish to build a set of binary files, which for a fresh install can take up to an hour of processing time.

## User manual

![NeuroElf GUI without datasets](_files/manual/NE_gui_empty.png)

### Preface

#### What NeuroElf is and what it's not

NeuroElf is a MATLAB toolbox supposed to facilitate a sub-set of neuro-functional imaging applications (preprocessing, regressions, visualization, post-hoc data extraction and analyses, etc.). The GUI (graphical user interface) that comes with NeuroElf is able to visualize and work on typically used voxel-based (and selected surface) file formats. The best analogy that comes to mind is that of a Swiss Army knife of neuro-functional imaging tools. That being said, it's important to understand what NeuroElf is not...

NeuroElf is, in itself, not a toolbox or program with which all aspects of imaging applications can be done. Most importantly, it does not come with a full set of preprocessing routines, mainly spatial normalization to a template space, and it thus requires at least one other major neuro-imaging application, such as SPM, FSL, or BrainVoyager QX to bring raw scanner data into a format and representation that allows group-based analyses. As a method of choice, SPM-based preprocessing can be easily configured and executed using a separate graphical interface in NeuroElf.

Generally speaking, the functionality implemented in NeuroElf can be understood as a set of tools that will facilitate and complement the (pre-) processing, as it is implemented in SPM5 or SPM8 (SPM12 support will be added shortly) as well as provide the user with relatively easy to script functions to access and visualize the resulting data. On its own, it allows to run either ordinary-least-squares or robust regressions on preprocessed data files, with a variety of options for nuisance regressors, parametric regression, (non-deconvolved and re-convolved) PPI, seed-correlations, and a relatively simple to use contrast manager together with beta extraction and time course plotting capabilities.

Most of the focus and effort in developing and implementing NeuroElf has been placed on making these processes as accessible and hassle-free as possible for the non-Computer-Science user, i.e. for psychologists.

Another large portion of features in NeuroElf relates to the visualization and ease of extracting data after an analysis has been run. To that end, NeuroElf features high-resolution slice-based montages--i.e. overlaying statistical information as colored heat-maps on anatomical data--, surface-based representation of data, and a 3D rendering engine. Additional features also include a beta plotter (more on that later), and a graphical interface visualizing stimulus onset-locked region-averaged time courses (event-related averaging).

In short, some of the main reasons to use NeuroElf are to simplify SPM-based preprocessing, to perform regression analyses, and to visualize and dig into the results of those.

Some additional tools, such as (single-level) mediation analyses, meta analyses, single-run ICA, linear classification, etc. are available, but not all of these have yet been fully implemented in the GUI, and as such may only be fully utilized by a more experienced MATLAB user. If you find a tool that you believe is suitable for inclusion in NeuroElf, please feel free to contact me with a request at info@neuroelf.net.

#### Acknowledgments

While most of the coding has been done by the main author, the following people deserve special mention for their countless contributions to the process of making and improving on NeuroElf:

Hedy Kober, who is now an Assistant Professor of Psychiatry and Psychology at Yale University, New Haven, CT. Hedy has, over the years, been extremely helpful by suggesting countless improvements in the user interface and in pointing out several bugs and issues. In addition, she has been a good friend and her very enthusiastic approach to NeuroElf and advocacy outside of Columbia University certainly kept it from ever becoming a "dead-end project".

Kevin Ochsner, who is a Professor of Psychology at Columbia University, New York, NY, has provided the means and, also funded by his research grants, most of the datasets on which NeuroElf's features were tested initially.

Additionally, thanks go to Cameron DeLeone, Federico DeMartino, Bryan Denny, Fabrizio Esposito, Andrew Gerber, Rainer Goebel, Armin Heinecke, Katie Insel, Zoran Josipovic, Jenna Reinen, Ajay Satpute, Jen Silvers, Jared van Snellenberg, and Noam Zerubavel, for being heavy users and contributors to the process of making NeuroElf a better version of itself over time, by finding bugs and sharing their preferences for working with functional imaging data, which makes the foundation of a user-friendly UI.

Some code has been loaned from other authors:

* Chih-Chung Chang and Chih-Jen Lin, the authors of LIBSVM
* Aapo Hyv&auml;rinen, the author of the FastICA algorithm
* Dirk-Jan Kroon, the author of the shear-warp rendering algorithm

And last but not least, I thank my partner, Changyu Zhu, who supports me at times when things seem a bit pointless, so I keep moving on anyway.

### Overview

The NeuroElf toolbox has three main components, a graphical user interface (`neuroelf_gui`), a file handling class (`@xff`) and a library of functions for general data processing (`@neuroelf`).

This readme (manual) uses the following notation: two greater-than symbols at the beginning of a paragraph (>>) in a "code box" indicate a single line of code to be entered at the MATLAB prompt (please enter without the >> symbols!). For a longer paragraph of code, an initial line (and possibly one or several subsequent lines) will begin with a per-cent symbol (%) indicating a MATLAB comment.

#### Installation

Depending on whether you downloaded the installation M-file+MAT-file version, or the flat folder version, or cloned from github, instructions slightly differ.

The archive containing NeuroElf is about 30 Megabytes in length. Once unpacked and installed, NeuroElf requires at least 100 Megabyte and, if all additional files are created, can require as much as 2.5 Gigabyte of disk space. So please make sure that the folder you choose as installation folder is on a medium that offers enough free space (i.e. does not come with a hard limitation or quota).

Please be aware that as a user of NeuroElf, you should have write-access to a few critical folders located in the `_files` folder, and also the files located in the `_core/config` folder. Without write access in these folders some functions may not work properly!

##### M-file version installation

To install NeuroElf, unzip the content (three files: `NeuroElf_v11.m`, `NeuroElf_v11.mat`, and `ne_eo.m`), then open MATLAB and change the current folder into the folder containing those three files. Then enter

```
>> NeuroElf_v11 -i
```

to start the installation process. You will be asked where to install the toolbox, with the default being the toolbox folder of your MATLAB installation. If you do not have write permissions in that folder, please choose "n" when being asked whether this is a good location and then enter a more suitable folder path. If in doubt, enter a path within your home folder (unless there is a quota restriction in place, at which point you should talk to your system administrator to determine a good location).

##### Flat folder version installation

Simply unzip the archive and move the unpacked files into a folder where you would like to install the toolbox. This should be a folder without spaces or other non alphanumeric characters in its full path and name.

##### NeuroElf setup function

After the files are successfully unpacked, the function `neuroelf_setup` must be called. If you installed using the M-file version, this will be done automatically. Otherwise, please enter

```
>> neuroelf_setup
```

at the MATLAB prompt manually. This function ensures that the toolbox has been correctly installed and is functional. At the end of this procedure, you will be asked whether `neuroelf_makefiles` (from the `@neuroelf` library) should be called as well. If you have also downloaded the auxiliary files (Colin brain, etc.; see below) and unpacked those into the `_files/colin` subfolder of the installation, you can safely say no. Otherwise, it is recommended that you let NeuroElf generate these additional files (which will take somewhere between 20 and 50 minutes, depending on your computer's speed and available free memory).

#### Folder structure

The installation folder has the following folder structure:

* `@neuroelf/` - class containing the function library within its private subfolder
* `@transimg/` - class used to blend 2D (RGB) images together
* `@transio/` - class allowing transparent IO file access (for large datasets)
* `@xff/` - class giving access to a variety of file formats (including functions/methods)
* `@xfigure/` - class providing object-oriented design patterns for to MATLAB's graphics handles
* `@xini/` - class providing an interface to *.ini configuration files
* `@xprogress/` - simplified class for progress bars
* `_core/` - control, text, and image files required for NeuroElf's core functionality
* `_files/` - auxiliary files required for additional functions

#### Files coming with NeuroElf

NeuroElf comes with a 1-mm resolution version of Colin Holmes' 27-scan average brain image, and a set of derived images and surface reconstructions is generated during the `neuroelf_makefiles` call, including 0.5-mm resolution versions for improved visualizations, sub-cortical surfaces, as well as flattened meshes and a head mesh for more complex surface scenes. These files must be located in the `_files/colin` subfolder of the installation. Some preliminary masks have also been derived from this dataset, and those are stored in the `_files/masks` subfolder.

In addition, NeuroElf contains a copy of the original Talairach labeled brain atlas, together with transformation files that relate this atlas approximately to ICBM/MNI space (derived using SPM8's segmentation on a ternary labeled version of the atlas). Files related to the Talairach atlas are located in the `_files/tal` subfolder.

Lastly, the `neuroelf_makefiles` function optionally downloads and installs a 1-mm resolution full-brain, anatomical-T1 DARTEL template from NeuroElf's website to allow using SPM8's DARTEL tools to warp any subject's brain (anatomical and functional data, if aligned) to a slightly better matching ICBM/MNI space. This template was derived from 305 anatomical scans collected at Columbia University over the course of 11 studies, and also included Colin Holmes' 27-scan average brain. These files will be stored in the `_files/spm/dartel/Template1mm` subfolder.

Further subfolders in the `_files` folder contain the cached information of the `@xff` class (which speeds up initializing NeuroElf considerably across MATLAB sessions); additional M-files in the `_files/contrib` folder that may be of interest for users who want to script NeuroElf; files downloaded from NeuroSynth (via the File -> NeuroSynth maps menu entry in the GUI) will be stored in `_files/neurosynth/terms`; SPM processing job specification files are stored in `_files/spm`; and a small set of predefined surface files (sphere, cylinder, cube) are located in `_files/srf`.

### The NeuroElf GUI

The GUI component can be started up by entering

```
>> neuroelf_gui
```

on the MATLAB prompt. Below is a screenshot of the GUI with some areas highlighted that will be explained in greater detail in this chapter.

![NeuroElf GUI with highlighted areas](_files/manual/NE_gui_areas.png)
###### NeuroElf GUI with highlighted areas: 1. Menu; 2. Object selector; 3a. Statistical map selector; 3b. Statistical map configuration; 4. Cluster/VOI tools; 5. Drawing tools; 6. Slice-based dataset view and browsing controls

#### GUI outline

The highlighted areas are:

1. Menu: used to configure NeuroElf (mostly under the File -> Options menu entry) and call functions or open additional UIs (e.g. for the GLM contrast manager or montage configuration).
2. File selector: used to switch between loaded objects, to set the orientation (translation, rotation, size) for objects, and to close (unload) the currently selected object--both for anatomical and statistical datasets.
3a. Statistical map selector: used to select one or multiple maps (in the same dataset); the button functions are

<table border=0 cellpadding=0><tr><td rowspan=9 height=196><img src="_files/manual/NE_stats_buttons.png" height=196 width=25 alt="NeuroElf GUI statistical map tool buttons" /></td>
<td height=22>move selected map(s) up the list</td></tr>
<tr><td height=20>move selected map(s) down the list</td></tr>
<tr><td height=20>inspect and/or change map properties (name, stats type, DF)</td></tr>
<tr><td height=20>delete selected map(s) from the container object </td></tr>
<tr><td height=16></td>
<tr><td height=20>compute a formula (expression) on map values</td></tr>
<tr><td height=20>sample selected maps into surface space (with selected surfaces)</td></tr>
<tr><td height=20>open GLM beta plotter (if GLM selected or referenced)</td></tr>
<tr><td height=20>select maps with same name across subjects (for GLMs)</td></tr>
</table>

3b. Statistical map configuration: used to configure a single statistical map (thresholds, visible tails, cluster size thresholding, coloring).

<table border=0><tr><td height=144><img src="_files/manual/NE_stats_buttons2.png" height=144 width=280 alt="NeuroElf GUI statistical map configuration" /></td>
<td width=420>The two edit boxes with values represent the lower and upper threshold for the currently selected (single) map. All statistical values (absolutely) below the lower boundary will not be shown (transparent). For LUT configuration, values will be scaled between 1 and the number of lookup-colors defined in the LUT and then be shown in that color. For RGB configuration, all values above the upper boundary will be shown in the colors configured with the "++" and "--" buttons (depending on the sign of the statistic and enabled tails). Values inbetween the boundaries will be shown in a color between the "+" and "++" color for the positive tail or between the "-" and "--" colors for the negative tail. The p-value dropdown can be used to set a specific p-value as the lower boundary. The p-value will be converted into the appropriate statistical threshold by NeuroElf. In addition to a "height threshold", a cluster size (extent) threshold can be set in the edit box following the "k-thresh:" label and enable with the checkbox right before.</td></table>

4. Cluster/VOI tools: used to select a cluster and update the cursor position to its peak coordinate, also possibly auto-extracting data from clusters. The buttons can be used to:

<table border=0 cellpadding=0><tr><td rowspan=10 height=224><img src="_files/manual/NE_VOI_buttons.png" height=224 width=25 alt="NeuroElf GUI VOI tool buttons" /></td>
<td height=22>highlight selected clusters in VMR</td></tr>
<tr><td height=20>restrict selected clusters to a sphere around the peak (or create spherical ROI)</td></tr>
<tr><td height=20>add VOI from Talairach atlas labels</td></tr>
<tr><td height=20>goto cluster peak closest to current browsing location</td></tr>
<tr><td height=20>zoom in on currently selected cluster</td></tr>
<tr><td height=20>cluster properties (not yet implemented)</td></tr>
<tr><td height=20>remove selected clusters</td></tr>
<tr><td height=20>open (or add) VOI file from disk to list of clusters</td></tr>
<tr><td height=20>save list of clusters into VOI file</td></tr>
<tr><td height=20>manually trigger data extraction from selected clusters (from GLM)</td></tr>
</table>

5. Drawing tools: used to configure the drawing pen and perform image operations.

<table border=0 cellpadding=0><tr><td rowspan=3 height=68><img src="_files/manual/NE_draw_buttons1.png" height=68 width=24 alt="NeuroElf GUI drawing tool buttons (1)" /></td>
<td height=22>disable drawing mode (set to simple browsing mode)</td></tr>
<tr><td height=20>configure and enable 2D (within-slice) drawing mode</td></tr>
<tr><td height=20>configure and enable 3D (volumetric) drawing mode</td></tr>
<tr><td rowspan=5 height=110><img src="_files/manual/NE_draw_buttons2.png" height=110 width=24 alt="NeuroElf GUI drawing tool buttons (2)" /></td>
<td height=22>perform flood-fill operation within value range from current position</td></tr>
<tr><td height=20>perform grow-region operation (extend marked voxels by 1 within range)</td></tr>
<tr><td height=20>smooth voxels within selected value range (and bounding box)</td></tr>
<tr><td height=20>smooth marking (incl. threshold and bounding box)</td></tr>
<tr><td height=20>smooth within detected borders (not yet implemented)</td></tr>
<tr><td height=20><img src="_files/manual/NE_draw_buttonsu.png" height=22 width=24 alt="NeuroElf GUI drawing tool buttons (undo)" /></td>
<td height=20>toggle between regular and "undo-" painting modes</td></tr>
<tr><td rowspan=5 height=110><img src="_files/manual/NE_draw_buttons3.png" height=110 width=24 alt="NeuroElf GUI drawing tool buttons (3)" /></td>
<td height=22>accept changes (to buffer)</td></tr>
<tr><td height=20>revert changes (from buffer)</td></tr>
<tr><td height=20>reload entire dataset from disk</td></tr>
<tr><td height=20>reload marked voxels (from buffer)</td></tr>
<tr><td height=20>reload all but marked voxels (from buffer)</td></tr>
<tr><td rowspan=3 height=66><img src="_files/manual/NE_draw_buttons4.png" height=66 width=24 alt="NeuroElf GUI drawing tool buttons (4)" /></td>
<td height=22>define ROI/VOI from marked voxels</td></tr>
<tr><td height=20>set underlay object (for enhanced drawing or registration check)</td></tr>
<tr><td height=20>toggle between regular and V16 data for VMR objects</td></tr>
</table>

Most functions that are invoked by the GUI (e.g. updating the slicing position, particularly when statistical maps are overlaid on voxel-based object) can be scripted by adding the sub-function and required options as arguments to the call to <span class="nefunction">neuroelf_gui</span>, such as in

```
>> neuroelf_gui('setslicepos', [-16, -24, -16]);
```

or

```
% set the surface viewpoint of satellite BS1a2b3c to 135-mark-45 degrees,
% [0, 0, 0] translation, and a 1.25 zoom factor at virtual time index 0
neuroelf_gui('setsurfpos', 'BS1a2b3c', {135, 45, [0, 0, 0], 1.25, 0});

% save a screenshot of that satellite into "surface_135_45_zoom125.png"
% using the high-quality settings
neuroelf_gui('screenshot', 'BS1a2b3c', 'surface_135_45_zoom125.png', 'high-q');
```

This allows to script functions ranging from data processing to visualization, also making it possible to create animation movies! The sections describing the main functionality of the GUI below contain the name and prototypical use of the underlying function and, where applicable, a brief example. If a function is, for example, implemented in the file `@neuroelf/private/ne_screenshot.m`, the function can be called with `neuroelf_gui('screenshot', ARGUMENTS{:});`.

To get help about a specific function, such as `screenshot`, use:

```
>> neuroelf_gui('help', 'screenshot')
```

#### GUI functions overview

Some of the main functions of the GUI (above and beyond the library of functions discussed in a later chapter) are

* loading most supported file formats into the viewer (Analyze/NIFTI files representing anatomical, time-course, and statistical data; as well as BrainVoyager QX file formats and AFNI's HEAD/BRIK format)
* loading BrainVoyager surfaces (SRF) into the surface view
* overlaying multiple statistical maps on anatomical data (as long as they are in the same object container, such as a multi-volume NIFTI or VMP)
* displaying RGB-coded data (multi-colored anatomical datasets, NIFTI)
* drawing on/into anatomical and RGB data
* create surface reconstructions from binary (foreground/background) images
* blending two anatomical datasets (to check registration or to allow painting in specific areas)
* displaying multiple surfaces together
* undocking a view into satellites (which are live and can be linked to the main viewer)
* morphing surfaces to a target intensity in a voxel-based image</li>
* smoothing surfaces (inflation, to-sphere morphs)
* morphing surfaces using a "virtual time axis"
* sampling statistical maps into surface space
* displaying time-courses at cursor for 4D objects (NIFTI, VTC)
* configuring multi-study (random- and fixed-effects) GLMs (MDM)
* computing contrasts/correlations for random- and fixed-effects GLMs
* computing instantaneous seed-correlation maps (VTC only)
* computing arbitrary formulas on existing maps (VMP only)
* running mediation analyses on RFX-GLMs (single-level)
* performing meta-analyses (using PLP format)
* loading fMRI quality check and motion parameters for visual inspection
* creating high-quality slice-, render-, and surface-based visualizations
* averaging anatomical files for a group-based dataset for stats overlay

##### Loading datasets

Use the File -> Open... menu entry (COMMAND/CTRL + o) to select any of the following formats to be loaded into the viewer:

* VMR (BrainVoyager's voxel-based MR datasets)
* HDR/NII (Analyze/NIftI datasets) --to load statistical data, please use File -> Open file as stats..., *.nii.gz supported
* HEAD (AFNI's HEAD+BRIK datasets)
* SRF (BrainVoyager's surface datasets) --also used for fiber-trackings
* VMP (BrainVoyager's voxel-based map datasets; extended content in .Map.RunTimeVars)
* SMP (BrainVoyager's surface-based map datasets; extended content in .Map.RunTimeVars)
* GLM (BrainVoyager's General Linear Model datasets; extended content in .RunTimeVars)
* VTC (BrainVoyager's voxel-based 4D time course datasets; extended content in .RunTimeVars)
* MSK (BrainVoyager's mask datasets)
* FMR (BrainVoyager's slice-based functional MR datasets; STC loaded automatically)
* AMR (BrainVoyager's anatomical-information MR datasets)
* MAP (BrainVoyager's slice-based map datasets; obsolete)
* DMR (BrainVoyager's diffusion MR datasets; DWI loaded automatically)

NIftI files that are gzip-compressed (*.nii.gz) will be decompressed in a temporary folder prior to being rendered.

Currently, only NIftI files can be used to store RGB data in a (X, Y, Z, 1, 3)-sized dataset. To convert a VMR into an RGB dataset use the VMR -> Export as RGB Nifti menu entry.

Loading datasets is coded in functions `@neuroelf/private/ne_openfile.m` and `@neuroelf/private/ne_openstatsfile.m`. To call this function in a script, use `neuroelf_gui('openfile', FILENAME)` or `neuroelf_gui('openstatsfile', FILENAME)` for HDR/NII files.

##### General browsing

To change the slicing position, simply click into the displayed (voxel-based) dataset at a new position. Alternatively, coordinates can be entered either as real-world (MNI or TAL, depending on dataset) coordinates or as (1-based) voxel indices into the current dataset. For datasets that contain 4D time series, the volume can be selected by either entering the number into the respecting edit box or by clicking into the displayed time course.

##### Overlaying maps

To overlay statistical data on an anatomical dataset, load both a VMR/HDR/NII/HEAD object (or use the packaged Colin dataset) as well as a statistical map object (VMP or HDR/NII with File -> Open file as stats...).

Once the statistical maps are loaded, the first list box in the left column of controls can be used to select one or multiple maps (if the format supports those).

If a single map is selected, the controls that configure the map (lower and upper threshold, positive/negative tail of a distribution, cluster size thresholding, and coloration settings) can be used.

The p-value dropdown control will not be updated when a map is changed and only functions as a way to set the thresholds (no read-out of currently set thresholds). If the cluster size thresholding is enabled and cluster size thresholds have been determined--either via alphasim for regular maps or as part of the simulation for meta-analysis maps--the p-value dropdown will also update the cluster size threshold accordingly.

Further thresholding options for VMP-based maps are available via the VMP -> Thresholding sub menu.

Slicing and overlaying data is coded in function `@neuroelf/private/ne_setslicepos.m` and for satellites in `@neuroelf/private/ne_setsatslicepos.m`. Scripting of this function (e.g. for screenshots) is possible via `neuroelf_gui('setslicepos', [X, Y, Z])` or `neuroelf_gui('setsatslicepos', 'BS1a2b3c', [X, Y, Z])`.

##### Drawing tools

The drawing tools can be invoked using either the regular pen icon (second from the top on the left side of the brain slicing) to paint only into the selected slice or by using the 3D pen icon (below), to paint in all three dimensions.

When the drawing is enabled for the first time (or another operation is performed that alters the dataset), an undo-buffer will be stored in the associated object's `.RunTimeVars.UndoBuffer` field, from which voxels that are to be recovered will be read.

The configuration dialog that pops up allows to set a radius size and shape (circle/sphere or square/cube) as well as a color code.


CHANGELOG.TXT
This file describes, in reverse order, important changes made to the toolbox.
For a more comprehensive list, please check out

http://neuroelf.net/wiki/doku.php?id=neuroelf_changelog

Changes made since version 1.0 (after September 14, 2014)
 - global signal can now be regressed out in the GLM beta plotter (invoked from a VMP)
 - robust weights for RFX-based t- and r-maps (contrast manager) can be stored in output VMP

Changes made since release v0.9d  (after March 13, 2014)
 - instantaneous VTC seed-correlation
 - SPM-based preprocessing now has DARTEL, file cleanup, and configuration saving/loading options
 - added the @neuroelf class as method interface (to reduce the number of visible M-files)
 - findfiles options can now be given as shorthands: f = findfiles(D, P, '-d1D');
 - all MEX files are now precompiled for currently supported platforms
 - libSVM was updated to v3.18
 - added SMP::ComputeFormula (also to GUI)
 - more flexible formatting of the cluster table output in the GUI (File -> Options)
 - added support for resizing the main GUI, and size changes are faster
 - the GUI now displays map type and DF in the VMP map list
 - tool to average anatomical scans completed
 - preliminary support for RSA added (non-normalized VTCs in GLMs, t-maps instead of betas)
 - option to export a VMR into an RGB-NIftI file for RGB drawing

Changes made since release v0.9c (after April 29, 2011)
 - completed the rendering engine with statistical overlay and RGB-volume rendering
 - added SVC (small-volume correction) to the GUI
 - making using of Chris Rorden's DICOM-to-NIftI conversion tool, dcm2nii, as an option
 - auto-detection of Unicode files for asciiread
 - if exactly two statistical maps are shown, a maximally-distance hue can be chosen for the overlap
 - better support for compressed NIftI files
 - importing SPM t-Maps also imports available RPV information for smoothness estimates
 - preprocessing GUI now gives the option to brain-extract the anatomical scans
 - full-brain onset-locked average time-courses can be created with MDM::VTCCondAverage
 - model-comparison function (modelcomp) to allow flexible F-testing of models
 - added support for Vince Calhoun's HRFboost method
 - computation of auto-regression (lagged) correlation in cov_nd
 - improvements for sub-clustering (taking the distance of voxels to the peak into account)
 - the contrast manager now also allows to compute maps "per group" instead of comparing groups
 - multiple-filetype selection in xff constructor with file = xff('*.(glm|vmp)'); pattern
 - added remote-control capabilities for the GUI via HTTP/webbrowser link
 - performance improvement for renamedicom function, read all first, then write out
 - MDM::ComputeGLM also supports 4D NIftIs direction (instead of VTCs)
 - new functions to estimate smoothness of data (mapestsmooth) and residuals (resestsmooth)
 - added fixed-effects support to MDM::ComputeGLM
 - MKDA analysis added to GUI (implementation in PLP::MKDA)
 - several keyboard shortcuts added to GLM beta plotter
 - new function allowing to concatenate partially matching structs (catstruct)
 - AutoSave field in RunTimeVars of objects added to ensure consistency of information
 - PRT (protocol) files now allow mulitple parameters per condition
 - preliminary support for BrainVoyager-style ANOVAs in mrmanova and GLM::Anova
 - new feature to "concatenate" series of 3D NIftIs with h = xff('path/pattern*.nii');

Changes made since release v0.9b (after June 15, 2010)
 - added support to display two anatomical datasets "overlaid" (underlay option)
 - exporting BrainVoyager formats into 4D NIftI (GLM, VMP, VTC, etc.)
 - option to plot a confidence ellipse over a beta plotter (scatter view)
 - VOI-based event-related averaging plots (MDM::VOICondAverage)
 - new function to convert activation-cluster MAT files (TW code) to VMPs, ac2vmp
 - conversion of regular to single-trial PRTs
 - added backwards compatibility with scripts for BVQXtools (using BVQXfile and isBVQXfile)
 - conversion from and to TAL is now done with icbm2tal and tal2icbm using DCT-based transform
 - implementation of Stouffer's Z added in function stoufferz
 - added (single-level) mediation GUI (implementation in mediationpset)
 - new function to regress data against as many models, mmregress
 - colorspace conversion function hsvconv added (MEX)
 - RunTimeVars now keeps track of discarded volumes for VTCs
 - added rendering capabilities to the GUI
 - neuroelf_gui function split into sub-functions in different files to simplify editing
 - new function, neuroelf_makefiles, to create a more feature-rich base dataset
 - added high-quality, manual segmentation of Colin brain to archive
 - faster marching-cubes algorithm for direct-border reconstruction of VMRs
 - function and methods to apply SPM's DCT-based normalization: applyspmsnc
 - new method to regress out confounds from VTC data, VTC::RegressOut
 - added GUI for computing MDM-based GLMs

Changes made since release v0.9a (first release, after May 17, 2010)
 - added "echo mode" in GUI to show what calls are made
 - any2ascii now supports the objects held in configurations (for @xini class)
 - fast and reliable implementation of reading text file tables in u8str2double (MEX)
 - support for physio data analysis (heart beats)
 - compact display of SPM-based motion parameters in showspmrparams



hxdouble.m
%HXDOUBLE  Transform a double into a 1x16 (or longer) ASCII string.
%   HXSTRING = HXDOUBLE(VALUE) converts the number(s) in VALUE to their
%   hexadecimal representation, such that they can be stored and later
%   converted back to numbers without loss of precision.
%
%   VALUE = HXDOUBLE(HXSTRING) converts the ASCII representation in
%   HXSTRING back to its double representation.
%
%   Importantly, for more than one value (vector, matrix, or array),
%   VALUE will be linearized with (:) first; hence for proper conversion
%   to the original VALUE, it is necessary to separately store the SIZE.
%
%   See also ANY2ASCII, SIZE, EVAL, HXSINGLE.

% Version:  v1.1
% Build:    16031110
% Date:     Mar-11 2016, 10:06 AM EST
% Author:   Jochen Weber, SCAN Unit, Columbia University, NYC, NY, USA
% URL/Info: http://neuroelf.net/

% Copyright (c) 2010 -2016, Jochen Weber
% All rights reserved.
%
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
%     * Redistributions of source code must retain the above copyright
%       notice, this list of conditions and the following disclaimer.
%     * Redistributions in binary form must reproduce the above copyright
%       notice, this list of conditions and the following disclaimer in the
%       documentation and/or other materials provided with the distribution.
%     * Neither the name of Columbia University nor the
%       names of its contributors may be used to endorse or promote products
%       derived from this software without specific prior written permission.
%
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
% ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
% WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
% DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
% DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
% (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
% ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
% (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

% function definition
function hxs = hxdouble(inp)

% persistent conversion array
persistent hxd_hxr;
if isempty(hxd_hxr)
    hxd_hxr = [ ...
        nan * ones(1, 48) ...
          0   1   2   3   4   5   6   7   8   9 nan nan nan nan nan nan ...
        nan  10  11  12  13  14  15 nan nan nan nan nan nan nan nan nan ...
        nan * ones(1, 16) ...
        nan  10  11  12  13  14  15 nan nan nan nan nan nan nan nan nan ...
        nan * ones(1,144)];
end

% number of arguments
if nargin ~= 1
    error('neuroelf:general:badNumberOfInputs', 'Invalid number of inputs.');
end

% for a double value/array
if isa(inp, 'double')

    % pass on to sprintfbx
    hxs = sprintfbx(inp);

% for a char
elseif ischar(inp)

    % linearize
    inp = inp(:)';

    % number of double values
    ns  = floor(numel(inp) / 16);

    % prepare output
    hxs = nan * zeros(1, ns);

    % iterate
    for hxc = 1:ns

        % get (lower-char) string
        ipp = lower(inp((hxc-1)*16+1:hxc*16));

        % NaN?
        if strcmp(ipp, 'fff8000000000000')
            continue;
        end

        % only allow ASCII, and add one for indexing into lookup
        ipd = min(double(ipp), 255) + 1;

        % lookup value
        ipi = hxd_hxr(ipd);

        % invalid string
        if any(isnan(ipi))
            continue;

        % all 0
        elseif all(ipi == 0)
            hxs(hxc) = 0;
            continue;
        end

        % compute exponent
        iex = 256 * ipi(1) + 16 * ipi(2) + ipi(3);

        % get sign correct
        if iex > 2047
            isg = -1;
            iex = iex - 2048;
        else
            isg = 1;
        end

        % compute mantissa
        imt = 0;
        for mtc = 4:16
            imt = imt*16+ipi(mtc);
        end

        % for numbers of regular magnitude
        if iex > 64
            hxs(hxc) = 2 .^ (iex-1023) + (2 .^ (iex-1075)) .* imt;

        % small numbers
        elseif iex > 0
            hxs(hxc) = 2 .^ (iex-1023) + (imt ./ (2.^63)) ./ (2.^1011);

        % tiny numbers
        else
            hxs(hxc) = (imt ./ (2.^63)) ./ (2.^1011);
        end

        % sign
        hxs(hxc) = hxs(hxc) * isg;
    end

% unsupported class
else
    error('neuroelf:hxdouble:invalidInputType', 'Invalid input type: %s.', class(inp));
end
